### PR TITLE
Checkpoint actions

### PR DESCRIPTION
**Pull request by mrinehart-r7**
### Description

Describe the proposed changes:

  - Adds actions: Check if Host in Group, Remove Host from Group

### Testing

Any testing information goes here.

## PR Requirements

Developers, verify you have completed the following items by checking them off:

### Style

Review the [style guide](https://komand.github.io/python/style.html)

- [ ] For dependencies, pin [OS package](https://komand.github.io/python/style.html#dockerfile) and [Python package](https://komand.github.io/python/style.html#requirements-txt) versions
- [ ] For security, set least privileged account with ``USER nobody`` in the ``Dockerfile`` when possible
- [ ] For size, use the [slim SDK images](https://komand.github.io/python/sdk.html#sdk-versions) when possible: ``komand/python-3-37-slim-plugin`` and ``komand/python-3-37-plugin``
- [ ] For error handling, use of [PluginException](https://komand.github.io/python/error_handling.html#plugin-exceptions) and [ConnectionTestException](https://komand.github.io/python/error_handling.html#connection-exceptions)
- [ ] For logging, use [self.logger](https://komand.github.io/python/sdk.html#logging)
- [ ] For docs, use [changelog style](https://komand.github.io/python/style.html#changelog)
- [ ] For docs, validate markdown with ``make validate`` which calls ``mdl`` to lint ``help.md``

### Functional Checklist
- [ ] Work fully completed
- [ ] Functional
  - [ ] Any new actions/triggers include JSON [test files](https://komand.github.io/python/style.html#tests) in the `tests/` directory created with `icon-plugin run -c sample $action > tests/$action.json`
  - [ ] Tests should all pass unless it's a negative test. Negative tests have a naming convention of `tests/$action_bad.json`
  - [ ] Unsuccessful tests should fail by raising an exception causing the plugin to die and an object should be returned on successful test
  - [ ] Add functioning test results to PR, sanitize any output if necessary
    * Single action/trigger `icon-plugin run -T tests/example.json --debug --jq`
    * All actions/triggers shortcut `icon-plugin run -T all --debug --jq` (use PR format at end)
  - [ ] Add functioning run results to PR, sanitize any output if necessary
    * Single action/trigger `icon-plugin run -R tests/example.json --debug --jq`
    * All actions/triggers shortcut `icon-plugin run -R all --debug --jq` (use PR format at end)

### Assessment

You must validate your work to reviewers:

1. Run `make validate` and make sure everything passes
2. Run the assessment tool: `icon-plugin run -A -R all -T all`. For single action validation: `icon-plugin run -A -R tests/my_action.json -T tests/my_action.json`
3. Copy (`icon-plugin ... | pbcopy`) and paste the output in **a new post** on this PR.
4. Add UI screenshot of the workflow used for testing
5. Add UI screenshot of the job output used for testing
6. Add UI screenshot of the artifact (See rules in UI Checklist) used for testing

### CODE REVIEW

File: checkpoint_ngfw/help.md [hunk 1]
```diff
@@ -27,26 +27,88 @@ The connection configuration accepts the following parameters:
...
   "port": 443,
-  "server": "192.168.1.1",
-  "ssl_verify": false,
-  "username_password": {
-    "password": "xxxxxx",
-    "username": "xxxxxx"
-  }
+  "server": "198.168.2.1",
+  "ssl_verify": true,
+  "username_password": "{\"username\": \"xxxxxx\", \"password\": \"xxxxxx\"}"
```
Comment (jmcadams-r7) [inline checkpoint_ngfw/help.md:39]: Fix this.

File: checkpoint_ngfw/help.md [hunk 2]
```diff
@@ -27,26 +27,88 @@ The connection configuration accepts the following parameters:
...
+#### Check if Address in Group
+
+This action checks to see if an IPv4 or IPv6 address is in an Address Group
+
+##### Input
+
+|Name|Type|Default|Required|Description|Enum|Example|
+|----|----|-------|--------|-----------|----|-------|
+|address|string|None|True|IPv4 or IPv6 address to check in the group|None|198.51.100.100|
+|discard_other_sessions|boolean|True|True|Discard all other user sessions. This can fix errors when objects are locked by other sessions|None|True|
```
Comment (jmcadams-r7) [inline checkpoint_ngfw/help.md:92]: Set this to False by default

File: checkpoint_ngfw/icon_checkpoint_ngfw/actions/check_if_address_in_group/action.py [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+        """
+        Returns all network groups found within the Check Point instance
+        :return: Network groups as a list of dictionaries
+        """
+        full_groups: [dict] = []
+
+        limit = 500
+        current_offset = 0
+
+        while True:
```
Comment (jmcadams-r7) [inline checkpoint_ngfw/icon_checkpoint_ngfw/actions/check_if_address_in_group/action.py:83]: This is style, so you can dissmiss this comment. 

I abhor "while True"

I'd do something like "while current_offset < total" to make it more clear what we're doing here and remove the break if possible. 

Dunno. It's not bad in the grand scheme of things, just not how I'd do it.

File: checkpoint_ngfw/icon_checkpoint_ngfw/actions/remove_address_object_from_group/action.py [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+                "remove": address_name
+            }
+        }
+
+        headers = self.connection.get_headers()
+        response = self.connection.post_and_publish(headers, discard_other_changes, payload, url)
+
+        if response.status_code == 200:
+            return {Output.SUCCESS: True}
+        elif response.status_code in [400, 401, 403, 404, 409, 500, 501]:
```
Comment (jmcadams-r7) [inline checkpoint_ngfw/icon_checkpoint_ngfw/actions/remove_address_object_from_group/action.py:35]: I think you should have a catch-all error at the end so there's not a possibility of them sending you a bad return code.

File: checkpoint_ngfw/plugin.spec.yaml [hunk 5]
```diff
@@ -885,3 +889,74 @@ actions:
...
+        required: true
+        example: 198.51.100.100
+        order: 2
+      discard_other_sessions:
+        title: Discard Other Sessions
+        type: boolean
+        description: Discard all other user sessions. This can fix errors when objects are locked by other sessions
+        example: true
+        required: true
+        default: true
```
Comment (jmcadams-r7) [inline checkpoint_ngfw/plugin.spec.yaml:917]:
```suggestion
        default: false
```

File: checkpoint_ngfw/plugin.spec.yaml [hunk 6]
```diff
@@ -885,3 +889,74 @@ actions:
...
+        example: ICON Block List
+        required: true
+        order: 2
+      discard_other_sessions:
+        title: Discard Other Sessions
+        type: boolean
+        description: Discard all other user sessions. This can fix errors when objects are locked by other sessions
+        example: true
+        required: true
+        default: true
```
Comment (jmcadams-r7) [inline checkpoint_ngfw/plugin.spec.yaml:955]:
```suggestion
        default: false
```

File: checkpoint_ngfw/help.md (hunk 1 above)
Comment (mrinehart-r7) [reply checkpoint_ngfw/help.md:39]: It's autogenerated by icon-plugin, it's going to get overwritten every time :(

File: checkpoint_ngfw/icon_checkpoint_ngfw/actions/remove_address_object_from_group/action.py (hunk 4 above)
Comment (mrinehart-r7) [reply checkpoint_ngfw/icon_checkpoint_ngfw/actions/remove_address_object_from_group/action.py:35]: Good idea

File: checkpoint_ngfw/help.md (hunk 1 above)
Comment (jschipp-r7) [reply checkpoint_ngfw/help.md:39]: that's what we have to do for accuracy

Comment (jschipp-r7) [reply checkpoint_ngfw/help.md:39]: it's not bad when you regen and do `git add -p` to selectively add changes

File: checkpoint_ngfw/help.md [hunk 7]
```diff
@@ -47,6 +49,67 @@ Example input:
...
+|----|----|-------|--------|-----------|----|-------|
+|address_object|string|None|True|The name of the address object (host) to remove|None|Malicious Host|
+|group|string|None|True|Group name|None|ICON Block List|
+
+Example input:
+
+```
+{
+  "address_object": "Malicious Host",
+  "group": "ICON Block List"
```
Comment (jschipp-r7) [inline checkpoint_ngfw/help.md:68]: ICON is mostly an internal name
```suggestion
  "group": "InsightConnect Block List"
```

File: checkpoint_ngfw/plugin.spec.yaml [hunk 8]
```diff
@@ -473,6 +474,13 @@ connection:
     required: true
     example: true
     default: true
+  discard_other_sessions:
+    title: Discard Other Sessions
+    type: boolean
+    description: Discard all other user sessions. This can fix errors when objects are locked by other sessions
```
Comment (jschipp-r7) [inline checkpoint_ngfw/plugin.spec.yaml:480]: I think this name could be more user friendly. 
This only happens on older versions of the firewall IIRC.
It's an API workaround, better to call it something like "Older than Rxx" but that name is still meh.

We should be more explicit here as well in the description and help.md on what this actually means.
discard other sessions translates to logging any existing users out of the Checkpoint UI, which means they could lose their sessions. We instituted this option as a workaround to an API bug in older versions of Checkpoint.

File: checkpoint_ngfw/plugin.spec.yaml [hunk 9]
```diff
@@ -738,19 +714,14 @@ actions:
         required: true
         type: string
         order: 1
+        example: "ICON Block List"
```
Comment (jschipp-r7) [inline checkpoint_ngfw/plugin.spec.yaml:717]:
```suggestion
        example: "InsightConnect Block List"
```

File: checkpoint_ngfw/plugin.spec.yaml (hunk 8 above)
Comment (jmcadams-r7) [reply checkpoint_ngfw/plugin.spec.yaml:480]: I disagree...I want to be very clear that we are discarding sessions here. I don't want to rely on the user reading the description (they won't) and accidentally blowing away someone else working on the server.

File: checkpoint_ngfw/help.md [hunk 10]
```diff
@@ -24,21 +24,23 @@ The connection configuration accepts the following parameters:
...
 
 Example input:
 
 ```
 {
+  "discard_other_sessions": true,
   "port": 443,
-  "server": "192.168.1.1",
-  "ssl_verify": false,
+  "server": "198.168.2.1",
```
Comment (jschipp-r7) [inline checkpoint_ngfw/help.md:39]: Use the IP address convention from doc

File: checkpoint_ngfw/help.md [hunk 11]
```diff
@@ -578,11 +629,15 @@ Example input:
...
-  "name": "Test from Komand",
-  "position": "top",
-  "source": "192.1.2.1"
+  "list_of_services": [
+    "AOL",
+    "SMTP"
+  ],
+  "name": "Malicious IP Addresses",
+  "position": 1,
+  "source": "192.168.2.1"
```
Comment (jschipp-r7) [inline checkpoint_ngfw/help.md:640]: use convention IP space

File: checkpoint_ngfw/help.md [hunk 12]
```diff
@@ -578,11 +629,15 @@ Example input:
 ```
 {
   "action": "Drop",
-  "discard_other_sessions": true,
+  "destination": "192.168.2.1",
```
Comment (jschipp-r7) [inline checkpoint_ngfw/help.md:632]: use convention IP space

File: checkpoint_ngfw/plugin.spec.yaml [hunk 13]
```diff
@@ -871,17 +847,64 @@ actions:
...
+        type: "[]string"
+        required: true
+
+  remove_address_object_from_group:
+    title: Remove Address Object from Group
+    description: Removes an address object (host) from an address group
+    input:
+      address_object:
+        title: Address Object
+        description: The name of the address object (host) to remove
```
Comment (jschipp-r7) [inline checkpoint_ngfw/plugin.spec.yaml:893]: I think they call they these host objects? if so, change all references in the spec and regen

```suggestion
        description: The name of the address object (host object) to remove
```

File: checkpoint_ngfw/plugin.spec.yaml [hunk 14]
```diff
@@ -728,29 +717,24 @@ actions:
...
         description: Name of the group to add this object to
         required: true
         type: string
         order: 1
-      host_name:
+        example: "InsightConnect Block List"
+      address_object:
         title: Host Name
-        description: The host to add to the network group, usually the IP address
+        description: The name of the host to add, usually the IP address
```
Comment (jschipp-r7) [inline checkpoint_ngfw/plugin.spec.yaml:733]: I'm wondering if `usually the IP address` will be confusing - might indicate that IPs are find to use (without them being address objects of the same name)

```suggestion
        description: The name of the host object to add
```

File: checkpoint_ngfw/plugin.spec.yaml (hunk 8 above)
Comment (jschipp-r7) [reply checkpoint_ngfw/plugin.spec.yaml:480]: I think the `Discard Other Sessions` is too technical a language to get the point across. When I first looked at it,  I didn't understand what was happening.

@mrinehart-r7's  suggestion of "Force Changes" was better - and we could auto-detect and only apply it if the version is older. So if they're on a new version where it's not needed it won't be used.

**Pull request merged**